### PR TITLE
KEP-4191: Split Image Filesystem promotion to Beta

### PR DESCRIPTION
**Pull request by kwilczynski**
- One-line PR description: Promote [KEP-4191](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/4191-split-image-filesystem/README.md?rgh-link-date=2024-06-03T08%3A18%3A51Z) "Split Image Filesystem" to Beta, update references to the `KubeletSeparateDiskGC` feature gate.

- Issue link: https://github.com/kubernetes/enhancements/issues/4191

- Other comments:

Starting from the upcoming Kubernetes release v1.31, the feature will be enabled by default, allowing users to deploy split image and container filesystems without enabling the feature gate to benefit from the new functionality.

Thus, update the documentation that references the `KubeletSeparateDiskGC` feature gate, ensuring that its state post-Beta promotion will be correctly reflected.

Related:

- https://github.com/kubernetes/website/pull/46951
- https://github.com/kubernetes/kubernetes/pull/126205

### CODE REVIEW

File: content/en/docs/reference/command-line-tools-reference/kube-apiserver.md
Comment (dipesh-rawat) [inline content/en/docs/reference/command-line-tools-reference/kube-apiserver.md:1]: Please exclude changes to the auto-generated [kube-apiserver reference page](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/#auto-generated-edit-notice). During the release process, this file's content is autogenerated (and then the generated docs are committed into this repo), so you don't need to amend them in this PR.

File: content/en/docs/reference/command-line-tools-reference/kube-controller-manager.md
Comment (dipesh-rawat) [inline content/en/docs/reference/command-line-tools-reference/kube-controller-manager.md:1]: Please exclude changes to the auto-generated [kube-controller-manager reference page](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/#auto-generated-edit-notice). During the release process, this file's content is autogenerated (and then the generated docs are committed into this repo), so you don't need to amend them in this PR.

File: content/en/docs/reference/command-line-tools-reference/kube-proxy.md
Comment (dipesh-rawat) [inline content/en/docs/reference/command-line-tools-reference/kube-proxy.md:1]: Please exclude changes to the auto-generated [kube-proxy reference page](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/#auto-generated-edit-notice). During the release process, this file's content is autogenerated (and then the generated docs are committed into this repo), so you don't need to amend them in this PR.

File: content/en/docs/reference/command-line-tools-reference/kube-scheduler.md
Comment (dipesh-rawat) [inline content/en/docs/reference/command-line-tools-reference/kube-scheduler.md:1]: Please exclude changes to the auto-generated [kube-scheduler reference page](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-scheduler/#auto-generated-edit-notice). During the release process, this file's content is autogenerated (and then the generated docs are committed into this repo), so you don't need to amend them in this PR.

File: content/en/docs/reference/command-line-tools-reference/kubelet.md
Comment (dipesh-rawat) [inline content/en/docs/reference/command-line-tools-reference/kubelet.md:1]: This file content is generated as part of the release process, and the generated docs are committed into this repository separately. Hence, there's no need to modify it within this pull request.

File: content/zh-cn/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md
Comment (dipesh-rawat) [inline content/zh-cn/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md:1]: Please exclude changes to this Chinese localization file.

File: content/zh-cn/docs/reference/command-line-tools-reference/kube-apiserver.md
Comment (dipesh-rawat) [inline content/zh-cn/docs/reference/command-line-tools-reference/kube-apiserver.md:1]: Please exclude changes to this Chinese localization file.

File: content/zh-cn/docs/reference/command-line-tools-reference/kube-controller-manager.md
Comment (dipesh-rawat) [inline content/zh-cn/docs/reference/command-line-tools-reference/kube-controller-manager.md:1]: Please exclude changes to this Chinese localization file.

File: content/zh-cn/docs/reference/command-line-tools-reference/kube-proxy.md
Comment (dipesh-rawat) [inline content/zh-cn/docs/reference/command-line-tools-reference/kube-proxy.md:1]: Please exclude changes to this Chinese localization file.

File: content/zh-cn/docs/reference/command-line-tools-reference/kube-scheduler.md
Comment (dipesh-rawat) [inline content/zh-cn/docs/reference/command-line-tools-reference/kube-scheduler.md:1]: Please exclude changes to this Chinese localization file.

File: content/zh-cn/docs/reference/command-line-tools-reference/kubelet.md
Comment (dipesh-rawat) [inline content/zh-cn/docs/reference/command-line-tools-reference/kubelet.md:1]: Please exclude changes to this Chinese localization file.

**Review by dipesh-rawat — COMMENTED** [review]
@kwilczynski It appears that this PR includes changes to files across multiple languages _English_ and _Chinese_. However, docs follows different processes for each localization, and we typically don't accept pull requests that impact multiple languages simultaneously. Also, there are changes to the files which have auto-generated contents. Those pages are not supposed to be manually updated.
 I would recommend modify only the primary website non-auto generated content (which means only English content) initially with this PR.

**Comment by dipesh-rawat** [comment]
Explicit hold until this PR only covers one language and excludes changes to auto-generated pages.
_(Ok to unhold once that's done)_

/hold

File: content/zh-cn/docs/reference/command-line-tools-reference/kubelet.md
Comment (kwilczynski) [reply content/zh-cn/docs/reference/command-line-tools-reference/kubelet.md:1]: Done.

File: content/zh-cn/docs/reference/command-line-tools-reference/kube-scheduler.md
Comment (kwilczynski) [reply content/zh-cn/docs/reference/command-line-tools-reference/kube-scheduler.md:1]: Done.

File: content/zh-cn/docs/reference/command-line-tools-reference/kube-proxy.md
Comment (kwilczynski) [reply content/zh-cn/docs/reference/command-line-tools-reference/kube-proxy.md:1]: Done.

File: content/zh-cn/docs/reference/command-line-tools-reference/kube-controller-manager.md
Comment (kwilczynski) [reply content/zh-cn/docs/reference/command-line-tools-reference/kube-controller-manager.md:1]: Done.

File: content/zh-cn/docs/reference/command-line-tools-reference/kube-apiserver.md
Comment (kwilczynski) [reply content/zh-cn/docs/reference/command-line-tools-reference/kube-apiserver.md:1]: Done.

File: content/zh-cn/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md
Comment (kwilczynski) [reply content/zh-cn/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md:1]: Done.

File: content/en/docs/reference/command-line-tools-reference/kubelet.md
Comment (kwilczynski) [reply content/en/docs/reference/command-line-tools-reference/kubelet.md:1]: Done.

File: content/en/docs/reference/command-line-tools-reference/kube-scheduler.md
Comment (kwilczynski) [reply content/en/docs/reference/command-line-tools-reference/kube-scheduler.md:1]: Done.

File: content/en/docs/reference/command-line-tools-reference/kube-proxy.md
Comment (kwilczynski) [reply content/en/docs/reference/command-line-tools-reference/kube-proxy.md:1]: Done.

File: content/en/docs/reference/command-line-tools-reference/kube-controller-manager.md
Comment (kwilczynski) [reply content/en/docs/reference/command-line-tools-reference/kube-controller-manager.md:1]: Done.

File: content/en/docs/reference/command-line-tools-reference/kube-apiserver.md
Comment (kwilczynski) [reply content/en/docs/reference/command-line-tools-reference/kube-apiserver.md:1]: Done.

**Comment by kwilczynski** [comment]
> Explicit hold until this PR only covers one language and excludes changes to auto-generated pages. _(Ok to unhold once that's done)_

@dipesh-rawat, done. Hopefully.

**Comment by kwilczynski** [comment]
> @kwilczynski It appears that this PR includes changes to files across multiple languages _English_ and _Chinese_. However, docs follows different processes for each localization, and we typically don't accept pull requests that impact multiple languages simultaneously. 
[...]

@dipesh-rawat, good to know! I had no idea... Hence, I went with every file in which I could find a reference to the feature gate.

**Comment by kwilczynski** [comment]
@dipesh-rawat, please have a look again, thank you!

**Comment by dipesh-rawat** [comment]
/unhold

PR now covers only single language.

File: content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md [hunk 1]
```diff
@@ -6,9 +6,13 @@ _build:
   render: false
 
 stages:
-  - stage: alpha
```
Comment (dipesh-rawat) [inline content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md:9]: _nit_: Avoid change to this existing line.

File: content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md [hunk 2]
```diff
@@ -6,9 +6,13 @@ _build:
   render: false
 
 stages:
-  - stage: alpha 
+  - stage: alpha
     defaultValue: false
     fromVersion: "1.29"
+  - stage: beta
+    defaultValue: true
+    fromVersion: "1.31"
```
Comment (dipesh-rawat) [inline content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md:14]: Adding `toVersion` to the previous stage would enhance consistency with other feature gate files.

```suggestion
    toVersion: "1.30"
  - stage: beta
    defaultValue: true
    fromVersion: "1.31"
```

**Review by dipesh-rawat — COMMENTED** [review]
@kwilczynski, We should consider setting the `dev-1.31` branch as the target branch rather than the _main_ branch for this pull request. Since the change is related to the feature targeting beta in 1.31.

**Comment by dipesh-rawat** [comment]
/hold

Explicit hold to avoid accidental merges to main branch.
_(Ok to unhold once this is targetted to dev-1.31)_

File: content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md [hunk 3]
```diff
@@ -6,9 +6,13 @@ _build:
   render: false
 
 stages:
-  - stage: alpha 
+  - stage: alpha
     defaultValue: false
     fromVersion: "1.29"
```
Comment (tengqm) [inline content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md:11]: add this after line 11:

```
  toVersion: "1.30"
```

File: content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md [hunk 4]
```diff
@@ -6,9 +6,13 @@ _build:
...
+  - stage: beta
+    defaultValue: true
+    fromVersion: "1.31"
 ---
-Enable kubelet to garbage collect container images and containers
-even when those are on a separate filesystem.
+The Split Image filesystem feature enables kubelet to perform garbage collection
```
Comment (tengqm) [inline content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md:16]:
```suggestion
The split image filesystem feature enables kubelet to perform garbage collection
```

File: content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md [hunk 5]
```diff
@@ -6,9 +6,13 @@ _build:
...
+of images (read-only layers) and/or containers (writeable layers) deployed on a
```
Comment (tengqm) [inline content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md:17]:
```suggestion
of images (read-only layers) and/or containers (writeable layers) deployed on
```

File: content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md (hunk 1 above)
Comment (kwilczynski) [reply content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md:9]: This line has surplus whitespaces that _should not be there_.

File: content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md (hunk 2 above)
Comment (kwilczynski) [reply content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md:14]: Done.

File: content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md (hunk 3 above)
Comment (kwilczynski) [reply content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md:11]: Done.

File: content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md (hunk 4 above)
Comment (kwilczynski) [reply content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md:16]: Done.

File: content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md (hunk 5 above)
Comment (kwilczynski) [reply content/en/docs/reference/command-line-tools-reference/feature-gates/kubelet-separate-disk-gc.md:17]: Done.

**Comment by kwilczynski** [comment]
> /hold
> 
> Explicit hold to avoid accidental merges to main branch. _(Ok to unhold once this is targetted to dev-1.31)_

@dipesh-rawat, done. Moved to the **dev-1.31** branch.

**Comment by dipesh-rawat** [comment]
/unhold

PR now targets `dev-1.31`

**Review by dipesh-rawat — COMMENTED** [review]
Thanks!

/lgtm

**Comment by tengqm** [comment]
@dipesh-rawat The upstream PR has not been merged ... the hold was not supposed to be lifted.

https://github.com/kubernetes/kubernetes/pull/126205

**Comment by dipesh-rawat** [comment]
> The upstream PR has not been merged ... the hold was not supposed to be lifted.

Apologies for the oversight on my part. I will raise a PR to revert the changes from dev-1.31 that were merged prematurely.

**Comment by sftim** [comment]
BTW, there's an option to document the behavior more, by editing:
- https://kubernetes.io/docs/concepts/architecture/garbage-collection/#image-maximum-age-gc
  Note that it's enabled by default
- https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#eviction-signals

**Pull request merged**